### PR TITLE
Add ContainerRuntimeSearchRegistries under RegistrySources

### PR DESCRIPTION
**Pull request by umohnani8**
Add the containerRuntimeSearchRegistries option to the
image.config.openshift.io CRD so that users can update
the list of registries to be searched when pulling using
short names.

Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

edit: bparees
api implementation of https://github.com/openshift/enhancements/pull/378
RFE: https://issues.redhat.com/browse/IR-142

### CODE REVIEW

**Comment by umohnani8** [comment]
@mrunalp PTAL

**Comment by umohnani8** [comment]
@sttts PTAL

**Comment by adambkaplan** [comment]
@umohnani8 is there an enhancement proposal to go along with this API change? This has impacts on builds, imagestream import, and the image registry (in addition to cri-o).

**Comment by umohnani8** [comment]
Yes, we have an enhancement here https://github.com/openshift/enhancements/blob/master/enhancements/api-review/add-unqualified-search-registries.md.
@mrunalp PTAL

**Comment by umohnani8** [comment]
@adambkaplan @abhinavdahiya can I please get a review here - the enhancement was accepted https://github.com/openshift/enhancements/blob/master/enhancements/api-review/add-unqualified-search-registries.md

File: config/v1/types_image.go [hunk 1]
```diff
@@ -112,4 +112,7 @@ type RegistrySources struct {
 	// Only one of BlockedRegistries or AllowedRegistries may be set.
 	// +optional
 	AllowedRegistries []string `json:"allowedRegistries,omitempty"`
+	// unqualifiedSearchRegistries are registries that will be searched when pulling images using short names.
```
Comment (abhinavdahiya) [inline config/v1/types_image.go:115]: since these doc comments end up as `oc explain` i would really to see 

- what are short names,
- can example of the setting and when we used with a short name what behavior is expected.
- comment on the order or priority wrt search order, are these in decreasing priority or container engine will decide that on its own
- how do i know if the setting has been applied to the cluster and i can use this feature now.
- any restrictions on these values like you can not have registries that are already set in blocked registries or can not set the registry that hosts the release image maybe etc..

Comment (abhinavdahiya) [reply config/v1/types_image.go:115]: nit: `unqualifiedSearchRegistries` for me this word has very little meaning i.e when seeing this field in documentation is is not quickly apparent to me what this would do.. just saying as someone who is not completely in know of the registry stuff.. maybe that is okay as you need that background context as you are editing the Image object.

Comment (adambkaplan) [reply config/v1/types_image.go:115]: @abhinavdahiya `unqualifiedSearchRegistries` corresponds directly to the configuration option in `registries.conf` [1] that is used by podman. I think it is fair to assume that a customer altering this option is familiar with the RHEL/podman equivalents.

With regard to "short names" - I suggest the following:

```
unqualifiedSearchRegistries are registries that will be searched when pulling images that do not have fully qualified domains in their pull specs. Registries will be searched in the order provided in the list.
```

[1] https://github.com/containers/image/blob/master/docs/containers-registries.conf.5.md

**Comment by adambkaplan** [comment]
Note that we already have a shared library to alter `registries.conf` in place: https://github.com/openshift/runtime-utils/blob/master/pkg/registries/registries.go#L104

**Comment by adambkaplan** [comment]
/hold

The EP states [1]:

> No new risks are introduced.

I am not convinced this is true.

From a technical perspective, the API for this configuration is well known and already in CRI-O. Applying this configuration via MCO or other means should be straightforward.

IIRC, however, there is quite a bit of special-case logic for `docker.io` in upstream kubernetes. Removing `docker.io` as a default search registry historically has been a source of major bugs and hard-to-track regressions. This is one of the reasons why builds hard-code `docker.io` in the search path today. Can we be assured that customers won't break their clusters or applications by altering this setting?

We must also bear in mind that 4.6 is an EUS release. Are we prepared to backport bugs for the next two years knowing that we are introducing what IMO is a risky feature?

Moreover, the enhancement proposal only covers changes at the node/CRI-O level. Customers altering the `image` config expect these changes to work across cluster features, which include:

1. Builds
2. Imagestream Import, and
3. Image registry pull through

Without a coordinated effort, customers will have a skewed experience where the unqualified search registries are used in some contexts (`Deployment`), but not in others (`Build`). This could have serious security implications if the customer vets images in their private registry, only to have users unknowingly bypass the private registry when they do a build or import an imagestream.

I may be missing other on-top components which do their own image pulls. Ex: is this a concern for OLM?

/cc @bparees @dmage @nstielau @shawn-hurley 

[1] https://github.com/openshift/enhancements/blob/master/enhancements/api-review/add-unqualified-search-registries.md#risks-and-mitigations
[2] https://github.com/openshift/builder/blob/master/imagecontent/etc/containers/registries.conf#L13-L14

**Comment by bparees** [comment]
Good list of concerns, @adambkaplan 

I agree that we do not want to introduce this api w/o a commitment to have all components that pull images/image metadata respect it (OLM does not today pull images directly, so that's not a concern).  We have only recently finally gotten ourselves consistent across the board on this, I would hate to lose it again.

I don't necessarily think there's a great risk of customers being broken by it.  I'm not sure exactly what behavior is being implemented, but if it's something like:

1) if the value is not set, use docker.io as the default search
2) if the value is set, use the searchlist provide by the value

Then that seems reasonable enough... no customer will be broken out the box (default behavior is not changing), and if they choose to set a value, that's on them (and they can include docker.io as the first entry in that list of values to avoid being broken)

I also recall that at one point in time, k8s itself actually prepended docker.io onto unqualified pod/container image pullspecs, before the container runtime ever saw the value, which would make this setting ineffective, but perhaps that is no longer the case?

**Comment by umohnani8** [comment]
The `unqualified-search-registries` field currently exists in the OCP 4.x clusters under `/etc/containers/registries.conf`. This is an option that was added when we moved to using v2 of containers/image - I am not sure why [1] is using the v1 format of registries.conf.

With this enhancement and API change, we are exposing this option so that the user can modify/override the list to include registries that they want searched when using image short names. The image search and pull are done by cri-o, podman, and buildah, so if the Builds, Imagestream Import, and Image registry pull components are not bypassing these tools to pull the image directly, the changes to the `Image` CR should work across all of these.

We plan to very clearly document how and when to use this. The only time for customers to do this is if they have internal registries that change their DNS often, which would require them to update the image spec every time that happens so that the image can be successfully pulled. The use of this option is going to be very rare. This is an effort to put all the Image configurations in one CR instead of having to create MC objects for some of them.

The MCO registries code and shared library registries code will need to be updated to account for this new API change. Are there any other components that I may have missed?

[1] https://github.com/openshift/builder/blob/master/imagecontent/etc/containers/registries.conf

**Comment by dmage** [comment]
ImageStream imports and the image-registry pull-through use Docker libraries and they are not aware of `registries.conf` (like CRI-O is not aware of `allowedRegistriesForImport` that exists since 3.9). So it's not that easy to teach these components about this new option. And I agree with @bparees, with shouldn't allow to change this until all components can support this option.

**Comment by bparees** [comment]
> The image search and pull are done by cri-o, podman, and buildah, so if the Builds, Imagestream Import, and Image registry pull components are not bypassing these tools to pull the image directly, the changes to the Image CR should work across all of these.

it's not a matter of "bypassing" crio, they (other than builds) have nothing to do w/ crio and have their own stack for interacting with registries given a pullspec because they are doing something different.

With respect to builds, they do make some effort to reconstruct a matching crio config for the build pod (which invokes buildah itself within its own environment, hence why the conf file must be reconstructed explicitly) that matches what the node uses, but that logic would need to be updated to include this behavior to ensure the conf files continue to match.  As noted by Oleg above, that's probably the less challenging aspect of ensuring all our components respect this configuration, however.

**Comment by adambkaplan** [comment]
@bparees does this enhancement also impact the image policy admission controller? I suspect not, trying to cover our bases.

**Comment by adambkaplan** [comment]
> The `unqualified-search-registries` field currently exists in the OCP 4.x clusters under `/etc/containers/registries.conf`. This is an option that was added when we moved to using v2 of containers/image - I am not sure why [1] is using the v1 format of registries.conf.
> 
> [1] https://github.com/openshift/builder/blob/master/imagecontent/etc/containers/registries.conf

This is likely a bug - I think we've been providing a default `registries.conf` since the 3.x days in the v1 format. Interesting, it looks like if allowed/blocked registries are set, then we generate a v2 formatted version of the file [2].

[2] https://github.com/openshift/runtime-utils/blob/master/pkg/registries/registries.go#L104

**Comment by bparees** [comment]
> @bparees does this enhancement also impact the image policy admission controller? I suspect not, trying to cover our bases.

i do not believe so since that operates at a level before any of this defaulting would come into play.

**Comment by umohnani8** [comment]
> ImageStream imports and the image-registry pull-through use Docker libraries and they are not aware of registries.conf (like CRI-O is not aware of allowedRegistriesForImport that exists since 3.9).

These components probably already handle insecure, blocked, and allowed registries - could we do the same for unqualified-search-registries?

For builds, we continue to make the effort to reconstruct the registries.conf that matches what the node has by adding logic to handle the `unqualified-search-registries` list. This way we will be consistent across the components that deal with images and registries.
I will update the enhancement with this information of components that will need updates to handle this API change.

**Comment by dmage** [comment]
@umohnani8 imports are going to support insecure, blocked, and allowed registries in 4.6. But we don't have support of these options for pull-through, and won't have in 4.6.

**Comment by umohnani8** [comment]
@dmage can imports add support for `unqualified-search-registries` as well in 4.6 then? And if pull-through doesn't have these options yet, that is something that will be implemented in the future so shouldn't be a a blocker for builds and MCO.
Also, it might be worth moving imports and pull-through to use the the containers/image library in the 4.7 time frame so we are consistent across all the components.

**Comment by dmage** [comment]
@umohnani8 I don't think the registry team has time to implement support of `unqualified-search-registries` in 4.6, there isn't much time before feature freeze. Also, it's unlikely that we'll migrate to `containers/image` in 4.7, we have higher priority features and I suspect we'll need to change `containers/image` before we can use it (in some cases it tries to interact via logrus, but apiserver and registry both are servers and users won't see logs; can it send additional headers? we use them to break circuits; it gets configuration via filesystem, perhaps we can live with it, but I don't like the idea that apiserver will write configs to files). Yet we try to use parts of `containers/image` whenever we can and migrating is in our backlog.

**Comment by umohnani8** [comment]
Good to know that the migration is in the backlog, thanks for the information @dmage. We are going to push this enhancement to 4.7 so that we have enough time to implement the API change in imports and registry-pull-through. I will leave this PR open till then and will follow up here once master opens for 4.7. I will add a story under the unqualified-search-registries EPIC for the registry team as well. Please let me know if there is anything else that I need to do :)

File: config/v1/types_image.go (hunk 1 above)
Comment (umohnani8) [reply config/v1/types_image.go]: Fixed with suggestion.

**Comment by bparees** [comment]
@umohnani8 since this recently came up on SME list as an issue in 3.11....  how does registry auth matching work when this is in place?

normally k8s (And other components) pick the registry creds to use based on the image pullspec that's referenced.  When there isn't one, how do those components pick the auth cred to use?  And then what happens if the search process picks a different registry than what the component assumed would be chosen?

**Comment by mtrmac** [comment]
> @umohnani8 since this recently came up on SME list as an issue in 3.11.... how does registry auth matching work when this is in place?
> 
> normally k8s (And other components) pick the registry creds to use based on the image pullspec that's referenced. When there isn't one, how do those components pick the auth cred to use? And then what happens if the search process picks a different registry than what the component assumed would be chosen?

tl;dr: it doesn’t work.

IIRC (warning: this is all from memory, not testing the current behavior) the Kubelet is completely unaware of the unqualified-search-registries configuration, and the CRI only allows passing down _one_ set of credentials (with a host name field that is ignored, i.e. without even specifying what the credentials are for). The outcome is that when the Kubelet encounters a short name, it will find the `docker.io` entry in the secret objects linked to the pull spec, which not only _fails_ when the search registries list uses different host names, it actively sends the `docker.io` credentials to a different registry!

The only way credentials can be provided to non-`docker.io` search registries is to include them in the CRI-O-visible default pull secret (and then they are available to all tenants of the cluster).

----

**I _very strongly_ recommend _against_ making `UnaulifiedSearchRegistries` user-controllable** any more than it currently is.
- The above credentials problem is not probably going to be fixed any time soon (it would require a CRI change)
- The search mechanism is _inherently_ insecure: When the user does not specify which registry to use, the registries, and network attackers, can cause an unintended image to be used (e.g. if `docker.io` is the first entry, `docker.io` can supply any image even if the user wanted to use an internal registry, or an attacker could break network connections to `docker.io` to cause a fallback to the second registry ).
- As a result, the search mechanism will be _significantly_ tightened up in future versions (at least in RHEL 9, if not sooner), to require an interactive confirmation of the intended registry in interactive situations, and …something else… in non-interactive ones. Expect the default to be “secure” eventually, i.e. at the very latest to refuse to operate on a search list with more than one entry, with other behaviors requiring opting in.

Users that want to do unsupported and unrecommended things can continue to do unsupported and unrecommended things by manually creating MachineConfigs. 

(I also consider the “our registry DNS names change very frequently” case to be extremely unconvincing.
- Why does it have to be this way? It just doesn’t make sense at a first glance.
- Fine, front the changing-DNS-name registry with a reverse proxy with a stable DNS name. That allows the customer to adapt to the DNS name change instantly and cheaply, instead of this feature causing MCO to re-deploy and re-boot the whole cluster!

)

---
 
A possible alternative, if the goal is only to support a rapidly-changing-DNS-name registry, might be to use ImageContentSourcePolicy (invent a stable DNS name, use it in pull secrets, and redirect it to the changing one). This doesn’t help if the user insists on using short names and not changing the deployment configuration.

OTOH ICSP currently only applies to digest references per the OpenShift-created configuration, to that would have to be relaxed (either by default, or with an opt-in).

---

If the search list _were_ a supported feature, the documentation/support status of that would have to be _very_ clear that:
- This subjects the users to network- and registry-originating attacks
- Credentials from pod pull secrets are not applied; only credentials from the cluster-wide pull secret are usable
- If the pod pull secrets contain `docker.io` credentials for any reason, they are going to be sent all registries in the search list.

**Comment by bparees** [comment]
thanks @mtrmac, it seems like we really need to revisit the EP, document these issues, and then decide if the EP still makes sense to move forward given the security+usability problems.

**Comment by umohnani8** [comment]
Thanks @mtrmac! After the discussion with @mtrmac and @mrunalp today, we believe the next step is to talk with the customers with the DNS name change scenario and see if using the ICSP would work for them - the drawback there right now is that it only works for images reference by digest. There is some discussions going on about enabling ICSP mirrors to work with tags and to make that an opt-in option, so if that goes through it should help with the ICSP use case for this scenario.

A major reason for exposing this in the Image CR instead of just leaving the user to use the MC work around is that only cri-o knows of the unqualified-search-registries list when using MCs and we wanted to spread this use case to all the components that do anything with image pulls. 

If there is a push to expose this option in the Image CR, then pod specs relying on this cannot have any pull secrets linked to it and will have to use the global pull secret that cri-o knows about. The customers are currently using the MC work-around to drop-in a file with their unqualified-search-registries list configuration and looks like that is working for them now. I will follow up on the exact use case and see if we can avoid exposing this option in the Image CR.
Agree on putting this on hold for now.

**Comment by umohnani8** [comment]
After a discussion with @bparees @mrunalp @dmage @mtrmac, we decided to move forward with exposing this option so that users can pull via short names. But we will only support this for the runtime case. We will not add support for this to builds and imagestream imports as there are security risks associated with the use of short names and we want to discourage it. Additionally, there is some complexity around figuring out the correct credentials/pull secret to pass down to the build and imagestream import components for authentication.

I updated the enhancement in https://github.com/openshift/enhancements/pull/537 to reflect the change in scope for this feature. I think we can get this moving along now, PTAL.

/hold cancel

File: config/v1/0000_10_config-operator_01_image.crd.yaml [hunk 2]
```diff
@@ -122,6 +122,16 @@ spec:
                   type: array
                   items:
                     type: string
+                unqualifiedSearchRegistries:
+                  description: 'unqualifiedSearchRegistries are registries that will
+                    be searched when pulling images that do not have fully qualified
+                    domains in their pull specs. Registries will be searched in the
+                    order provided in the list. Note: this search list does NOT work
+                    with builds and imagestream imports. It only works with the container
+                    runtime, CRI-O.'
```
Comment (bparees) [inline config/v1/0000_10_config-operator_01_image.crd.yaml:131]: i wonder if we want to pick a different name here, because if we ever did want to support search for imagestreams+builds, we would have to add a new field.  We couldn't just suddenly change how builds+imagestreams do shortname resolution.

If we named this field in a way that made it clear it only applied to crio pulls, then if we wanted to add build+imagestream support in the future, we could introduce this field name for that, taking precedence over the old field.

does that make sense?

Comment (umohnani8) [reply config/v1/0000_10_config-operator_01_image.crd.yaml:131]: yup makes sense. Something like `RuntimeShortNameRegistries` or `RuntimeSearchRegistries` or `RuntimeShortNamePullRegistries`? If not runtime we could use crio, wdyt?

Comment (bparees) [reply config/v1/0000_10_config-operator_01_image.crd.yaml:131]: RuntimeSearchRegistries could work or maybe NodeSearchRegistries since it is affecting image specs consumed by the node?

Comment (umohnani8) [reply config/v1/0000_10_config-operator_01_image.crd.yaml:131]: Both sound good to me, no particular preference. @mrunalp wdyt?

Comment (mrunalp) [reply config/v1/0000_10_config-operator_01_image.crd.yaml:131]: How about `ContainerRuntimeSearchRegistries`?

Comment (bparees) [reply config/v1/0000_10_config-operator_01_image.crd.yaml:131]: > How about ContainerRuntimeSearchRegistries?

i like it

Comment (umohnani8) [reply config/v1/0000_10_config-operator_01_image.crd.yaml]: Fixed! PTAL :)

**Comment by bparees** [comment]
thanks for all the discussions, @umohnani8

File: config/v1/types_image.go [hunk 3]
```diff
@@ -112,4 +112,9 @@ type RegistrySources struct {
 	// Only one of BlockedRegistries or AllowedRegistries may be set.
 	// +optional
 	AllowedRegistries []string `json:"allowedRegistries,omitempty"`
+	// containerRuntimeSearchRegistries are registries that will be searched when pulling images that do not have fully qualified
+	// domains in their pull specs. Registries will be searched in the order provided in the list.
+	// Note: this search list only works with the container runtime, i.e CRI-O. Will NOT work with builds or imagestream imports.
+	// +optional
+	ContainerRuntimeSearchRegistries []string `json:"containerRuntimeSearchRegistries,omitempty"`
```
Comment (sttts) [inline config/v1/types_image.go:119]: which format do these have? Please add a specification (OpenAPI pattern, format, minLength, unique; https://book.kubebuilder.io/reference/markers/crd-validation.html).

Are these unique? Can there be duplicates? Is this a set? Compare https://book.kubebuilder.io/reference/markers/crd-processing.html?highlight=listtype#crd-processing `ListType`.

Comment (umohnani8) [reply config/v1/types_image.go:119]: There are no specific length or format requirements here. There can be duplicates in the list. This list already exists on the node in `/etc/containers/registries.conf`, exposing this helps the user configure the list on the node. This new option is similar to the existing options under `RegistrySources` currently.
Do we still need to add specifications in this case?

Comment (sttts) [reply config/v1/types_image.go:119]: Yes, as part of a spec these are supposed to be fully specified. Adding restriction later is hard. So we have to be precise from the beginning.

I am pretty sure these cannot be empty string. And they are URLs? And this looks like a set.

Comment (umohnani8) [reply config/v1/types_image.go:122]: Got it, added the restrictions that I believe should apply to this field. PTAL

Comment (umohnani8) [reply config/v1/types_image.go:122]: @sttts is this good now, PTAL

**Comment by sttts** [comment]
LGTM schema-wise.

/assign @adambkaplan @bparees
